oscore-cf/src/main/java/org/eclipse/californium/oscore/CoapOSExceptionHandler.java

```diff
@@ -0,0 +1,117 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.coap.CoAP.Type;
+import org.eclipse.californium.core.coap.EmptyMessage;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+
+/**
+ * 
+ * Handles OSExceptions (determines if an EmptyMessage should be sent) and
+ * CoapOSExceptions (creates and returns the response with the correct response
+ * code).
+ *
+ */
+public class CoapOSExceptionHandler {
+
+	/**
+	 * The logger
+	 */
+	private static final Logger LOGGER = LoggerFactory.getLogger(CoapOSExceptionHandler.class.getName());
+
+	/**
+	 * 
+	 * Process the exception e which was provoked by the received request.
+	 * 
+	 * @param e the exception which was thrown because of the received request.
+	 * @param request the received request.
+	 * @return the appropriate error message in a response.
+	 * @throws OSException
+	 */
+	public static Response manageError(CoapOSException e, Request request) throws OSException {
+		ResponseCode responseCode = e.getResponseCode();
+		Response error = null;
+		boolean includeErrMess = true;
+
+		if (request != null) {
+			if (request.getType() != null) {
+				if (ResponseCode.isClientError(responseCode)) {
```

Comment:
What should be the intended behaviour for "isServerError"?

Reply:
Stop processing.

Reply:
Without ACK or RST? Then the coap client will usually repeat its request (at least, if it's a CON).
Just to mention, coap uses two number ranges, one indicating, that the client has to change, the other indicating, that the server is in an error state and the client message maybe processed at a other time. So I don't see, why only client errors should be forwarded. Maybe, currently only client errors are used, but if this is extended in the future, I think, server errors should be sent as well.

Reply:
In the case of isServerError a different function should be used all together, which in some cases replies with an EmptyMessage. See line 98 in CoapOSExceptionHandler